### ISSUE TITLE
Fibonacci heap infinite loop in consolidate() method

### ISSUE THREAD

**Issue by Flamak** [jgrapht/jgrapht#339]
Hi,

I have a few graph search algorithms while re-using the Fibonacci heap implementation in JGraphT and I have run into an infinite loop within consolidation method of the heap, in particular:

`while (x != minNode) {
                numRoots++;
                x = x.right;
            }`

Since x == x.right, the loop runs infinitely.

**Comment by d-michail** [comment]
@jsichi I actually came across the same bug yesterday in a different context. 

However, it is not that isolated. Somehow the list of roots gets corrupted during the heap operations.
It would help if you had any simple reproducible example.

**Comment by besza** [comment]
Shouldn't the FibonacciHeapNode class have a well-defined equals/hashCode? There are several object reference comparisons in the FibonacciHeap class. Could this lead to undefined behaviour?

**Comment by jsichi** [comment]
@besza  Hmmm, well, the original introduction of Nathan Fiedler's FibonacciHeap implementation was for use in ClosestFirstIterator; when I added it there, I intended it to be only used in contexts where we were already keeping a map from values (e.g. vertices) to heap nodes.  This ensures that multiple nodes never exist for the same value, meaning it is safe to use node object reference comparisons.  But I didn't document the assumption, so if newer algorithms have overlooked this, then yes, that could certainly be the cause of undefined behavior.

**Comment by d-michail** [comment]
This assumption can be easily enforced by not letting the user create the node before the insertion. 

This would mean:
 - making the constructor of `FibonacciHeapNode` package-private (or the node an inner class with private constructor) and 
 - changing method `insert` of `FibonacciHeap` into 
``` public FibonacciHeapNode insert(double key, T data) ```
  which would create and return the newly created node.

This is what I have done in an unrelated implementation of a Fibonacci Heap.

**Comment by d-michail** [comment]
@jsichi, @jkinable Actually this bug manifested in our newest A* implementation which also supports inconsistent heuristics (see pull-request #336). The reason is that we remove using `removeMin` a node from the heap and then re-insert it if necessary with a new key. 

The bug is that method `removeMin` forgets to initialize the removed node's internal state and method `insert` assumes that the node already represents a circular linked list. Adding the following code 
```
// clear z
z.left = z;
z.right = z;
z.degree = 0;
z.child = null;
z.mark = false;
```
in method `removeMin()` solves it.

I have added the fix in pull-request #317 which will be ready for merging in a couple of days.

@Flamak, @besza, @jsichi Thanks for the report and the discussion. They saved me a lot of debugging hours!

**Comment by jkinable** [comment]
@jsichi since you are the original author of `FibonacciHeapNode`, could you review the suggested bugfix for the FibonacciHeapNode in pull request #317? Aside from that, pull request #317 is almost finished (at least, I've got nothing left to complain about :) ). Feel free to request any changes. If not, then I'll merge #317

**Comment by jsichi** [comment]
The removeMin improvement from @d-michail looks fine to me.

**Comment by d-michail** [comment]
I opened up pull-request #340 with a slightly more generic fix which throws an `IllegalArgumentException` if a user tries to insert a node into a heap while the node already belongs to the same or some other heap. It is allowed, however, to remove a node from a heap and then insert it into another.

This should cover all cases that involve misusing a single heap.

As noted there, you can still misuse the implementation by using more than one heaps and performing operations on a node from another heap. Such cases require more advanced solutions (keeping on each node a reference to the heap using union-find due to possible cascaded union operations, etc.) and are out of scope.

**Comment by jsichi** [comment]
I've merged the PR from @d-michail.  Note that the map from values to nodes remains the responsibility of the caller.

**Comment by jkinable** [comment]
This issue has been resolved in pull req #317 and #340. Thanks for reporting! Let us know if you encounter any other issues.
